### DISCUSSION TITLE
Cannot get RequestPurchaseAsync to display anything.

### DISCUSSION THREAD

**Discussion by StrAbZ** [microsoft/WindowsAppSDK#3597] [Q&A]
Hi All!

Before opening an issue (because there might not be), I want to see if I'm missing something.

I have a app which implements un unlimited `Trial` and I cannot get `RequestPurchaseAsync` to do display the store window, either calling it on the `StoreContext` or a `StoreProduct`.

It instantly returns with the `Status` `notPurchased` and the `ExtendedError` to `null`

My app is associated with the store, I can retrieve the Product and License, so I can display a Dialog with the product price etc...
But the purchase process never "starts"

Here is the latest version of my code calling it:

```
(App.Current as App).Window.DispatcherQueue.TryEnqueue(() => {
    var  context = StoreContext.GetDefault();
    var hWnd = WinRT.Interop.WindowNative.GetWindowHandle((App.Current as App).Window);
    WinRT.Interop.InitializeWithWindow.Initialize(context, hWnd);
    var storePurchaseResult = await context.RequestPurchaseAsync(product.StoreId);
});
```

Previously the context was initialised way earlier, for debug purpose I put everything in the same call (but this changes nothing).

For information:
- The app was an UWP app, which I migrated to Windows App SDK
- The app belonged to another Partner account and only had 1 submission to be able to retrieve the License and test
- The app was changed to my current Partner account and was re-submitted to be able to retrieve the License and test
- In both case it was deployed but hidden, only available with direct link
- I cannot get to install it with another account on my machine, with the direct store link I get the error 0x80073CFB (and the message saying there was an error on MS Store side).
- I don't have access to a full Windows machine, so I'm working in a VM on a Mac M1 (not sure if arm64 stuff can have impact).
- I'n the VM, I have the last version of Windows 11 and the App projet is configured with:
```
    <TargetFramework>net6.0-windows10.0.22621.0</TargetFramework>
    <TargetPlatformMinVersion>10.0.22621.0</TargetPlatformMinVersion>
```

Thank you for any feedbacks or ideas you could have!

**Comment by duracellko** [comment]
Are you getting any exception? Possibly you are getting exception: `System.Runtime.InteropServices.COMException (0x80070578): Invalid window handle. (0x80070578)` Is that your case?

In my case I found this error, even though `InitializeWithWindow` is made. I use the same call for `MessageDialog` and that is displayed correctly. So I don't think the window initialization is actual problem. However, I found out that changing the application to full-trust RequestPurchaseAsync displays the purchase window. So it seems it does not work correctly in partial trust.

Specifically, when having following in `Package.appxmanifest`

```xml
<Applications>
  <Application Id="App"
    Executable="$targetnametoken$.exe"
    EntryPoint="Windows.PartialTrustApplication">
  </Application>
</Applications>
```

then the `PurchaseRequestAsync` does not work.

And when having following setup

```xml
<Applications>
  <Application Id="App"
    Executable="$targetnametoken$.exe"
    EntryPoint="$targetentrypoint$">
  </Application>
</Applications>
```

then it works correctly.

Could anyone provide a hint, how to make MS Store purchase work in partial trust application?

**Comment by DarranRowe** [comment]
Honestly, the experience for using UWP isn't great in desktop applications. This has an interesting case where there is some [documentation](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-supported-api#methods-that-use-the-request-naming-pattern) stating that it is outright unavailable, but the documentation for the runtime class itself states IInitializeWithWindow works.

I wonder if the underlying problem with this and partial trust application is the same as with the file pickers. It detects that the application is running in an app container and just assumes that it is a UWP application and CoreWindow exists. If that is true then these are things that really need fixing in Windows.

**Reply by duracellko** [reply]
You are right. I did some debugging and the RequestPurchaseAsync calls IsAppContainer and based on that accesses CoreApplication and CoreWindow or not.

**Comment by StrAbZ** [comment]
The application is already Full Trust, so I don't think it's that.

For now I open the Windows Store app to the product page so user can still buy "from the app", not perfect, but better than nothing.

Unfortunately, I purchase a version myself, and now, my app is no longer in Trial, so the `notPurchased` result might come from this, and I don't know how to "revoke" the licence so I could test again as a user who did not purchase it.

**Comment by duracellko** [comment]
How do you open Windows Store? Is there a URI that opens Windows Store app with purchase of certain in-app product?

**Reply by StrAbZ** [reply]
I use this: `await Windows.System.Launcher.LaunchUriAsync(new Uri("ms-windows-store://pdp/?productid="+product.StoreId));`

I don’t know if the `productid` can be something else than an app ID

**Comment by duracellko** [comment]
Thank you very much. Unfortunately, it does not work (not found) with in-app products.

**Comment by alski** [comment]
Just to add in, with exactly the same problem.

Again, application is FullTrust
```
 <trustInfo xmlns="urn:schemas-microsoft-com:asm.v2">
   <security>
     <requestedPrivileges xmlns="urn:schemas-microsoft-com:asm.v3">
       <requestedExecutionLevel level="requireAdministrator" uiAccess="false" />
     </requestedPrivileges>
   </security>
 </trustInfo>
```

and calling either of 

```
var result = await context.RequestPurchaseAsync(addOn.StoreId);
//var result = await addOn.StoreProduct.RequestPurchaseAsync();
``` 

leads to 

`WinRT information: This function must be called from a UI thread` 

while I am ![image](https://github.com/user-attachments/assets/534342e4-0ec5-4625-beb5-d9bfed3795c5)

**Reply by bogdan-patraucean** [reply]
Your issue is that you did not use `InitializeWithWindow`, and, if you are running your app as administrator, it will simply not work, thanks Microsoft...

**Comment by hth26** [comment]
StoreContext storeContext = StoreContext.GetDefault();
//for WinUI 3
//hWnd = WinRT.Interop.WindowNative.GetWindowHandle(this);
//WinRT.Interop.InitializeWithWindow.Initialize(storeContext, hWnd);
var result = await storeContext.RequestPurchaseAsync("9PE6X58NMWF9"); //Store-ID -> not "MyAddOn"
